File: 4.8-bc/py/stage-2-feature-first/app/features/notifications/handler.py
from __future__ import annotations

from datetime import datetime
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel
from sqlalchemy.ext.asyncio import AsyncSession

from app.features.notifications.repository import Repository
from app.features.notifications.service import Service
from app.shared.db import get_session

router = APIRouter(tags=["notifications"])


class SendTestReq(BaseModel):
    user_id: str
    channel: str | None = None
    payload: str | None = None


class SendTestResp(BaseModel):
    notification_id: str
    channel: str
    sent_at: datetime | None


@router.post("/notifications/test", response_model=SendTestResp)
async def send_test(
    req: SendTestReq, session: AsyncSession = Depends(get_session)
) -> SendTestResp:
    try:
        uid = UUID(req.user_id)
    except (ValueError, TypeError) as exc:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST, detail="invalid_user_id"
        ) from exc
    channel = req.channel or "email"
    payload = req.payload or "test notification"
    svc = Service(Repository(session))
    n = await svc.send(uid, channel, payload)
    return SendTestResp(
        notification_id=str(n.id), channel=n.channel, sent_at=n.sent_at
    )

File: 4.8-bc/ts/stage-2-feature-first/src/features/notifications/handler.ts
import type { FastifyInstance } from "fastify";

import { writeError } from "../../shared/httputil.js";
import type { Service } from "./service.js";

interface SendTestBody {
  user_id?: string;
  channel?: string;
  payload?: string;
}

const UUID_RE =
  /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;

export class Handler {
  constructor(private readonly svc: Service) {}

  register(app: FastifyInstance): void {
    app.post<{ Body: SendTestBody }>(
      "/notifications/test",
      async (req, reply) => {
        const body = req.body ?? {};
        if (!body.user_id || !UUID_RE.test(body.user_id)) {
          return reply.code(400).send({ error: "invalid_user_id" });
        }
        const channel = body.channel || "email";
        const payload = body.payload || "test notification";
        try {
          const n = await this.svc.send(body.user_id, channel, payload);
          return reply.code(200).send({
            notification_id: n.id,
            channel: n.channel,
            sent_at: n.sentAt,
          });
        } catch (err) {
          req.log.error({ err }, "send notification failed");
          return writeError(reply, err);
        }
      },
    );
  }
}
